src/api/views.py

```diff
@@ -1,73 +1,98 @@
     EventSerializer,
     UriSerializer,
 )
 
+
 bp = Blueprint('api', __name__, url_prefix='/api')
 
 
-class UriViewSet(MethodView):
-    """ API endpoint to display URIs. """
+class TokensViewSet(MethodView):
+    """ API endpoint to get JWT a API token. """
 
-    # permission_classes = (IsAuthenticated,)  # TODO
+    def post(self):
+        request_data = request.get_json()
+        user = User.query.filter_by(
+            email=request_data.get('email'),
+        ).first()
 
-    serializer_class = UriSerializer
+        password = request_data.get('password')
+        if not (user and user.verify_and_update_password(password)):
+            abort(401, "Invalid user credentials")
 
-    def get(self):
-        schema = self.serializer_class(many=True, only=('raw', 'last_checked'))
-        uri_data, errors = schema.dump(Uri.query.all())
-        return json.dumps(uri_data)
+        return issue_token(email=user.email)
 
-    def post(self):
 
-        """
+bp.add_url_rule('/get_token', view_func=TokensViewSet.as_view('get-token'))
+
 
-        !!! Need to authenticate user first and link their account to this
-        upload and doi.
+class UriViewSet(MethodView):
+    """ API endpoint to post and display URIs. """
 
-        Also, check if the doi exists and add user to this doi if that is the
-        case.
+    decorators = [token_authenticated]
 
-        """
+    serializer_class = UriSerializer
 
-        """New expected format of DOIs:
+    def get(self):
+        schema = self.serializer_class(
+            many=True,
+            only=('raw', 'urls', 'last_checked')
+        )
+        uri_data, errors = schema.dump(
+            Uri.query.filter(Uri.users.contains(g.user))
+        )
+        return uri_data
+
+    def post(self):
+        """ Register DOIs in the Altmetrics service. Also check if the URLs
+        provided, as well as the user making the request, need to be associated
+        with the DOI if it already exists.
+
+        Expected format of DOIs:
             [
                 {
-                    doi: '10.123.xxxxx'
-                    url: ['www.xxx.com', www.yyy.com],
+                    doi: '10.123/xxxxx'
+                    url: ['www.site.com/booklink', 'www.press.co.za/read'],
                 },
                 {
-                    doi: '10.456.aaaaa'
-                    url: ['www.aaa.com', www.bbb.com],
+                    doi: '10.456/aaaaa'
+                    url: ['www.site.org/book/doi', 'www.books.gov.uk/read/doi'],
                 },
                 ... etc
             ]
         """
-        json_input = request.get_json()
-
-        doi_list = json_input.get("doi_list")
+        doi_list = request.get_json()
 
         for doi_dict in doi_list:
-            raw_uri = doi_dict['doi']
-            new_uri = Uri(raw=raw_uri, last_checked=None)
-            db.session.add(new_uri)
+
+            uri = Uri.query.filter(Uri.raw == doi_dict['doi']).first()
```

Comment:
Yes, I was wondering about this. `'doi'` will definitely need to be in the `doi_dict` since this is how we identify an entry.
I think I'll add a `try/except` to give a meaningful error when this key is missing.